README.md
# Alfresco in Docker

## Images created (not all available for all versions)

* [`alfresco-skeleton`] = skeleton common for all "core" images per major version; includes java, tomcat, init script, keystore
* [`alfresco-enterprise`] = enterprise Alfresco-only images
* [`alfresco-community`] = community Alfresco-only images
* [`alfresco-enterprise-bundle`] = bundle images, with Alfresco+Solr+Postgres+JodConverter

## Supported Tags

* [`:5.1.3`, `:5.1`, `:5`, `:5.2.g`] = minor, major, revision for enterprise + community
* [`-bundle:4.2.6`,`-bundle:4.2`,`-bundle:4`] = minor, major, revision for bundled images
* [`-bundle:4.2.6-onbuild`] = not build automatically; bundled images including amps

## Overview

This is our repository with _Dockerfile_ s that are the cannonical source for our Alfresco Docker images. 

Images are automatically built by [jenkins-2](https://jenkins-2.xenit.eu) and published to [hub.xenit.eu](https://hub.xenit.eu).

## Image Variants

There are a few image variants, with different use cases in mind.

* Multi-container: `alfresco` to be used together with [`docker-solr`](https://bitbucket.org/xenit/docker-solr), [`postgres`](https://bitbucket.org/xenit/docker-postgres), [`jodconverter-ws`](https://bitbucket.org/xenit/docker-jodconverter-ws)
* Bundle: `alfresco-enterprise-bundle` contains a single image with _Alfresco_, _solr_, _postgresql_, _libreoffice_
* Onbuild bundle: `alfresco-enterprise-bundle:<version>-onbuild` extends from `-bundle` to quickly create a containerized version of Alfresco with your custom .amp

### alfresco-enterprise:<version>

This is the defacto core image for a multi-container deployment, especially in production.

You can use the [`src/main/resources/docker-compose.yml`](https://bitbucket.org/xenit/docker-alfresco/src/master/src/main/resources/docker-compose.yml) or [`src/main/resources/docker-compose-solr4.yml`](https://bitbucket.org/xenit/docker-alfresco/src/master/src/main/resources/docker-compose-solr4.yml) as a template in your project. 

Note that this uses the compose file [version 3](https://docs.docker.com/compose/compose-file/#version-3) syntax, which means you need [Docker Compose 1.10.0+](https://github.com/docker/compose/releases?after=1.10.1).

Images can be customized further by using environment variables - see section Environment Variables.

### alfresco-enterprise-bundle:4.2

If you need a quick disposable vanilla Alfresco image, this will suit you fine.

```
docker run --name my-alfresco -p 8080:8080 hub.xenit.eu/alfresco-enterprise-bundle:4.2
```

If you want your data to be safe, you can use a named volume:

```
docker run --name my-alfresco -p 8080:8080 -v my-alf-data:/data hub.xenit.eu/alfresco-enterprise-bundle:4.2
```

See the [`enterprise-4.2/bundle/docker-compose.yml`](https://bitbucket.org/xenit/docker-alfresco/src/master/4.2/enterprise-4.2.6-bundle/docker-compose.yml) for an example to get up and running.

### alfresco-enterprise-bundle:4.2-onbuild

TODO document onbuild with example

## Environment variables

There are several environment variables available to tweak the behaviour. While none of the variables are required, they may significantly aid you in using these images.
The variables are read by an init script which further replaces them in the relevant files. Such relevant files include - for alfresco:

* alfresco-global.properties
* server.xml (ports)
* context.xml (persistent sessions)
* setenv.sh (JAVA_OPTS parameters)

The alfresco-global.properties can be set via a generic mechanism by setting environment variables of the form GLOBAL_<parameter>, e.g. GLOBAL_alfresco.host. 
They can also be set via environment variables of the form JAVA_OPTS_<ignored_key> where the value should be "-Dkey=value".

A subset of the alfresco-global.properties have also dedicated environment variables e.g. SOLR_SSL. Generic variables take precedence.

A subset of java variables have also dedicated environment variables e.g. JAVA_XMX. Generic variables take precedence.

Environment variables:

| Variable                    | alfresco-global.property variable | java variable                                                | Default                                                      | Comments |
| --------------------------- | --------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------------- |
| SOLR_SSL                    | solr.secureComms                  |                                                              | https                                                        | disabling only works for Alfresco>=5.1 |
| ALFRESCO_HOST               | alfresco.host                     |                                                              | localhost                                                    |  |
| ALFRESCO_PORT               | alfresco.port                     |                                                              | 8080                                                         |  |
| ALFRESCO_PROTOCOL           | alfresco.protocol                 |                                                              | http                                                         |  |
| SHARE_HOST                  | share.host                        |                                                              | localhost                                                    |  |
| SHARE_PORT                  | share.port                        |                                                              | 8080                                                         |  |
| SHARE_PROTOCOL              | share.protocol                    |                                                              | http                                                         |  |
| DB_DRIVER                   | db.driver                         |                                                              | org.postgresql.Driver                                        |  |
| DB_HOST                     | db.host                           |                                                              | localhost                                                    |  |
| DB_PORT                     | db.port                           |                                                              | 5432                                                         |  |
| DB_NAME                     | db.name                           |                                                              | alfresco                                                     |  |
| DB_USERNAME                 | db.username                       |                                                              | alfresco                                                     |  |
| DB_PASSWORD                 | db.password                       |                                                              | admin                                                        |  |
| DB_URL                      | db.url                            |                                                              | jdbc:postgresql://postgresql:5432/alfresco                   |  |
| DB_QUERY                    | db.pool.validate.query            |                                                              | select 1                                                     |  |
| INDEX                       | index.subsystem.name              |                                                              | solr for alfresco 4 <br>solr4 for alfresco 5<br>solr6 for alfresco >=5.2 |  |
|                             | solr.host                         |                                                              | solr                                                         |  |
| SOLR_PORT                   | solr.port                         |                                                              | 8080                                                         |  |
| SOLR_PORT_SSL               | solr.port.ssl                     |                                                              | 8443                                                         |  |
| DYNAMIC_SHARD_REGISTRATION  | solr.useDynamicShardRegistration  |                                                              | false                                                        |  |
| MAIL_HOST                   | mail.host                         |                                                              | localhost                                                    |  |
| ENABLE_CIFS                 | cifs.enabled                      |                                                              | false                                                        |  |
| ENABLE_FTP                  | ftp.enabled                       |                                                              | false                                                        |  |
| ENABLE_CLUSTERING           | alfresco.cluster.enabled          |                                                              | false                                                        |  |
| TOMCAT_PORT                 |                                   | -DTOMCAT_PORT                                                | 8080                                                         |  |
| TOMCAT_PORT_SSL             |                                   | -DTOMCAT_PORT_SSL                                            | 8443                                                         |  |
| TOMCAT_AJP_PORT             |                                   | -DTOMCAT_AJP_PORT                                            | 8009                                                         |  |
| TOMCAT_SERVER_PORT          |                                   | -DTOMCAT_SERVER_PORT                                         | 8005                                                         |  |
| TOMCAT_MAX_HTTP_HEADER_SIZE |                                   | -DTOMCAT_MAX_HTTP_HEADER_SIZE  or -DMAX_HTTP_HEADER_SIZE                              | 32768                                                        |  |
| TOMCAT_MAX_THREADS          |                                   | -DTOMCAT_MAX_THREADS or -DMAX_THREADS                                              | 200                                                          |  |
| JAVA_XMS                    |                                   | -Xmx                                                         |                                                              |  |
| JAVA_XMX                    |                                   | -Xms                                                         |                                                              |  |
| DEBUG                       |                                   | -Xdebug -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=n |     false                                                         |  |
| JMX_ENABLED                 |                                   | -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.rmi.port=5000 -Dcom.sun.management.jmxremote.port=5000 -Djava.rmi.server.hostname=$JMX_RMI_HOST |     false                                                         |  |
| JMX_RMI_HOST                |                                   |                                                              |  0.0.0.0                                                            |  |
| GLOBAL_\<variable\>           | \<variable\>                        |                                                              |                                                              |  |
| JAVA_OPTS_\<variable\>=\<value\>       |                                   | \<value\>                                                   |                                                              |  |

## Docker-compose files

Besides the docker-compose files used in the tests, there are other example files in [src/main/resources](https://bitbucket.org/xenit/docker-alfresco/src/master/src/main/resources/).

## Support & Collaboration

Issues can be reported on the [JIRA DOCKER](https://xenitsupport.jira.com/projects/DOCKER) project.

These images are updated via pull requests to the [xenit/docker-alfresco/](https://bitbucket.org/xenit/docker-alfresco/) BitBucket-repository.


## FAQ

### How do I access the Tomcat debugport ?

Set the environment variable DEBUG=true. The debug port is 8000.

### How do I enable JMX?

Set the environment variable JMX_ENABLED=true. Jmx port is 5000.




Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v0.0.6] - 2018-10-26
### Added
* [DOCKER-126] Image for community 6.0.7-ga

### Removed
* [DOCKER-128] Remove community image 6.0.0, is unstable


## [v0.0.5] - 2018-09-24
### Changed
* [DOCKER-115] Move healthchecks to Dockerfiles

### Fixed
* [DOCKER-91] Redirect port was not parametrized in server.xml, therefore it was not possible to change it

## [v0.0.4] - 2018-09-10
### Added
* [DOCKER-76] Smoke tests

### Fixed
* [DOCKER-107] Property solr.port was not set due to wrong check in the init.sh

### Removed
* [DOCKER-120] Removed ALFRESCO_PORT and ALFRESCO_PORT_SSL referring to ports in server.xml. 
Correct way to change those - via TOMCAT_PORT and TOMCAT_PORT_SSL.


## [v0.0.3] - 2018-08-28
### Added
* [DOCKER-93] Image for Alfresco Enterprise 6.0
* [DOCKER-102] Image for Alfresco Enterprise 5.2.4 (last SP)

### Fixed
* [DOCKER-101] Fix tag on Alfresco Community 5.2.g

## [v0.0.2] - 2018-07-25
### Changed
* [DOCKER-98] Refactorings: simplified and deduplicated build.gradle, addition of "local" resources 

### Security
* [DOCKER-96] Do not echo properties being replaced, they should not appear in the logs (some are confidential)
	
## [v0.0.1] - 2018-07-19
### Changed
* [DOCKER-82] Good defaults for properties
* [DOCKER-12] [DOCKER-80] [DOCKER-79] [DOCKER-61] [DOCKER-55] [DOCKER-28] Refactorings: created global resources (for dockerfiles - with arguments, keystore, init), enterprise + community specific properties, single build.gradle, better namings
* [DOCKER-66] Adapted solr images names
* [DOCKER-31] [DOCKER-38] Tomcat-specific variables implemented as JAVA_OPTS and renamed into TOMCAT_<variable> (e.g. TOMCAT_PORT, TOMCAT_MAX_THREADS).
* [DOCKER-37] Adapt bundle images for new structure

### Added
* [DOCKER-92] Image for Alfresco Enterprise 5.2.3
* [DOCKER-87] Image for Alfresco Enterprise 4.2.8
* [DOCKER-67] Include jod converter amp
* [DOCKER-57] Added alfresco pdf renderer executable to Alfresco images >=5.2
* [DOCKER-94] Declare growing folders as volumes
* [DOCKER-35] Support for JAVA_OPTS_<variable> variables, allowing for overrides in different docker-compose files  
* [DOCKER-43] First version Changelog

### Removed
* [DOCKER-37] Removed bundled images for versions >=5.0
* [DOCKER-26] Removed the PROXY_<variable> parameters from the init
	



src/test/java/eu/xenit/docker/alfresco/test/AlfrescoSmokeTests.java
package eu.xenit.docker.alfresco.test;

import io.restassured.RestAssured;
import io.restassured.authentication.PreemptiveBasicAuthScheme;
import io.restassured.parsing.Parser;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.restassured.RestAssured.given;
import static io.restassured.http.ContentType.JSON;
import static java.lang.Thread.sleep;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.equalToIgnoringCase;
import static org.junit.Assert.*;

public class AlfrescoSmokeTests {

    @BeforeClass
    public static void setup() {
        String basePath = "/alfresco/s/";
        String host = System.getProperty("alfresco.host");
        int port = Integer.parseInt(System.getProperty("alfresco.tcp.8080"));

        RestAssured.baseURI = "http://" + host;
        RestAssured.port = port;
        RestAssured.basePath = basePath;

        PreemptiveBasicAuthScheme authScheme = new PreemptiveBasicAuthScheme();
        authScheme.setUserName("admin");
        authScheme.setPassword("admin");
        RestAssured.authentication = authScheme;
        RestAssured.defaultParser = Parser.JSON;
    }


    @Test
    public void testUp(){
        String versionRegex = "^" + System.getProperty("version") + ".+$";
        String flavor = System.getProperty("flavor");
        String version = given()
                .when()
                .get("api/server")
                .then()
                .statusCode(200)
                .contentType(JSON)
                .body("data.version", RegexMatcher.matchesRegex(versionRegex))
                .extract().path("data.version");
        String edition = given()
                .when()
                .get("api/server")
                .then()
                .statusCode(200)
                .contentType(JSON)
                .extract().path("data.edition");
        assertThat(edition.trim(), equalToIgnoringCase(flavor));
    }

    @Test
    // Note: due to eventual consistency, we can't test the expected number of results
    public void testSearch(){
        given()
                .when()
                .get("slingshot/search?term=milestone*")
                .then()
                .statusCode(200);
    }
}


src/main/resources/global/init.sh
#!/bin/bash

# we should get env ${CATALINA_HOME} from upstream container docker
set -e

JAVA_XMS=${JAVA_XMS:-'2048M'}
JAVA_XMX=${JAVA_XMX:-'2048M'}
DEBUG=${DEBUG:-'false'}
JMX_ENABLED=${JMX_ENABLED:-'false'}
JMX_RMI_HOST=${JMX_RMI_HOST:-'0.0.0.0'}

TOMCAT_ROOT=${CATALINA_HOME}
CONFIG_FILE=${TOMCAT_ROOT}'/shared/classes/alfresco-global.properties'
TOMCAT_CONFIG_FILE=${TOMCAT_ROOT}'/bin/setenv.sh'
TOMCAT_SERVER_FILE=${TOMCAT_ROOT}'/conf/server.xml'
TOMCAT_CONTEXT_FILE=${TOMCAT_ROOT}'/conf/context.xml'

SOLR_SSL=${SOLR_SSL:-'https'}

# sets an Alfresco parameter in alfresco-global.properties
function setGlobalOption {
    if grep --quiet -e "$1\s*=" "$CONFIG_FILE"; then
        # replace option
        sed -i "s#^\($1\s*=\s*\).*\$#\1$2#" $CONFIG_FILE
        if (( $? )); then
            echo "setGlobalOption failed (replacing option $1=$2 in $CONFIG_FILE)"
            exit 1
        fi
    else
        # add option if it does not exist
        echo "$1=$2" >> $CONFIG_FILE
    fi
}

# sets an Alfresco / Tomcat parameter as a JAVA_OPTS parameter
# the key is ignored, the value should contain the "-D" flag if it's a property
function setJavaOption {
    JAVA_OPTS="$JAVA_OPTS $2"
}

function setGlobalOptions {
    IFS=$'\n'
    for i in `env`
    do
	if [[ $i == GLOBAL_* ]]
	    then
	    key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 2-`
	    value=`echo $i | cut -d '=' -f 2-`
	    setGlobalOption $key $value
	fi
    done
}

function setJavaOptions {
    IFS=$'\n'
    for i in `env`
    do
	if [[ $i == JAVA_OPTS_* ]]
	    then
	    key=`echo $i | cut -d '=' -f 1 | cut -d '_' -f 3-`
	    value=`echo $i | cut -d '=' -f 2-`
	    setJavaOption $key $value
	fi
    done
}

setGlobalOption 'alfresco.host' "${ALFRESCO_HOST:-localhost}"
setGlobalOption 'alfresco.port' "${ALFRESCO_PORT:-8080}"
setGlobalOption 'alfresco.protocol' "${ALFRESCO_PROTOCOL:-http}"

setGlobalOption 'share.host' "${SHARE_HOST:-localhost}"
setGlobalOption 'share.port' "${SHARE_PORT:-8080}"
setGlobalOption 'share.protocol' "${SHARE_PROTOCOL:-http}"

setGlobalOption 'db.driver' "${DB_DRIVER:-org.postgresql.Driver}"
setGlobalOption 'db.host' "${DB_HOST:-localhost}"
setGlobalOption 'db.port' "${DB_PORT:-5432}"
setGlobalOption 'db.name' "${DB_NAME:-alfresco}"
setGlobalOption 'db.username' "${DB_USERNAME:-alfresco}"
setGlobalOption 'db.password' "${DB_PASSWORD:-admin}"
setGlobalOption 'db.url' "${DB_URL:-jdbc:postgresql://postgresql:5432/alfresco}"
setGlobalOption 'db.pool.validate.query' "${DB_QUERY:-select 1}"

# Search
if [[ $ALFRESCO_VERSION = "4"* ]]  # version 4 uses the bundled Alfresco with solr
then
    setGlobalOption 'index.subsystem.name' "${INDEX:-solr}"
elif [[ $ALFRESCO_VERSION = "5.0"* ]] || [[ $ALFRESCO_VERSION = "5.1"* ]]
then
    setGlobalOption 'index.subsystem.name' "${INDEX:-solr4}"
elif [[ $ALFRESCO_VERSION = "5.2"* ]]
then
    setGlobalOption 'index.subsystem.name' "${INDEX:-solr6}"
else
    setGlobalOption 'index.subsystem.name' "${INDEX:-solr6}"
fi

setGlobalOption 'solr.host' "${SOLR_HOST:-solr}"
setGlobalOption 'solr.port' "${SOLR_PORT:-8080}"
setGlobalOption 'solr.port.ssl' "${SOLR_PORT_SSL:-8443}"
setGlobalOption 'solr.useDynamicShardRegistration' "${DYNAMIC_SHARD_REGISTRATION:-false}"
setGlobalOption 'solr.secureComms' "${SOLR_SSL:-https}"
if [[ $SOLR_SSL = none ]] && [[ $ALFRESCO_VERSION != "5.0"* ]] && [[ $ALFRESCO_VERSION != "3"* ]] && [[ $ALFRESCO_VERSION != "4"* ]]
then
#remove the SSL connector
sed -i '/<Connector port="\${TOMCAT_PORT_SSL}" URIEncoding="UTF-8" protocol="org.apache.coyote.http11.Http11Protocol" SSLEnabled="true"/,+5d' $TOMCAT_SERVER_FILE
fi

# System
setGlobalOption 'mail.host' "${MAIL_HOST:-localhost}"
setGlobalOption 'cifs.enabled' "${ENABLE_CIFS:-false}" # CIFS Configuration
setGlobalOption 'ftp.enabled' "${ENABLE_FTP:-false}" # CIFS Configuration
setGlobalOption 'alfresco.cluster.enabled' "${ENABLE_CLUSTERING:-false}" # Cluster Configuration

# Tomcat-related properties
# for backwards compatibility, we keep the old options as well
setJavaOption 'TOMCAT_PORT' '-DTOMCAT_PORT='${TOMCAT_PORT:-8080}
setJavaOption 'TOMCAT_PORT_SSL' '-DTOMCAT_PORT_SSL='${TOMCAT_PORT_SSL:-8443}
setJavaOption 'TOMCAT_AJP_PORT' '-DTOMCAT_AJP_PORT='${TOMCAT_AJP_PORT:-8009}
setJavaOption 'TOMCAT_SERVER_PORT' '-DTOMCAT_SERVER_PORT='${TOMCAT_SERVER_PORT:-8005}
BC_TOMCAT_MAX_HTTP_HEADER_SIZE=${TOMCAT_MAX_HTTP_HEADER_SIZE:-$MAX_HTTP_HEADER_SIZE}
setJavaOption 'TOMCAT_MAX_HTTP_HEADER_SIZE' '-DTOMCAT_MAX_HTTP_HEADER_SIZE='${BC_TOMCAT_MAX_HTTP_HEADER_SIZE:-32768}
BC_TOMCAT_MAX_THREADS=${TOMCAT_MAX_THREADS:-$MAX_THREADS}
setJavaOption 'TOMCAT_MAX_THREADS' '-DTOMCAT_MAX_THREADS='${BC_TOMCAT_MAX_THREADS:-200}

if [ $JMX_ENABLED = true ]
then
    JAVA_OPTS="$JAVA_OPTS -Xms$JAVA_XMS -Xmx$JAVA_XMX -Dfile.encoding=UTF-8 -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.rmi.port=5000 -Dcom.sun.management.jmxremote.port=5000 -Djava.rmi.server.hostname=$JMX_RMI_HOST"
    setGlobalOption 'alfresco.jmx.connector.enabled' 'true'
else
    JAVA_OPTS="$JAVA_OPTS -Xms$JAVA_XMS -Xmx$JAVA_XMX -Dfile.encoding=UTF-8"
fi

if [ $DEBUG = true ]
then
    JAVA_OPTS="$JAVA_OPTS -Xdebug -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=n"
fi

sed -i "s|\(<Context>\)|\1\n<Manager pathname=\"\" />|" ${TOMCAT_CONTEXT_FILE} || exit 1 # No persistent sessions after restart

setGlobalOptions
setJavaOptions

echo "JAVA_OPTS=\""\$JAVA_OPTS" $JAVA_OPTS\"" >$TOMCAT_CONFIG_FILE
echo "export JAVA_OPTS" >> $TOMCAT_CONFIG_FILE

appStart () {
    if [[ $ALFRESCO_FLAVOR == enterprise-bundle ]]
    then
	user="alfresco"
    else
	user="tomcat"
    fi
    if [[ $(stat -c %U /opt/alfresco/alf_data) != "$user" ]]
    then
	chown -R $user:$user /opt/alfresco/alf_data
    fi
    if [[ $(stat -c %U "$CATALINA_HOME/temp") != "$user" ]]
    then
	chown -R $user:$user "$CATALINA_HOME"/temp
    fi
    if [[ $ALFRESCO_FLAVOR == enterprise-bundle ]]
    then
	gosu "$user" /opt/alfresco/alfresco.sh start
	tail -f /opt/alfresco/tomcat/logs/catalina.out
    else
	gosu "$user" ${CATALINA_HOME}/bin/catalina.sh run
    fi
}

appHelp () {
  echo "Available options:"
  echo " app:start    - Start the application"
  echo " app:help     - Displays the help"
  echo " [command]    - Execute the specified linux command eg. bash."
}


case "$1" in
  app:start)
    appStart
    ;;
  app:help)
    appHelp
    ;;
  *)
    if [ -x $1 ]; then
      $1
    else
      prog=$(which $1)
      if [ -n "${prog}" ] ; then
        shift 1
        $prog $@
      else
        appHelp
      fi
    fi
    ;;
esac

exit 0
